octoprint_authldap/__init__.py

```diff
@@ -0,0 +1,271 @@
+
+	_localUserManager = FilebasedUserManager()
+
+	def __init__(self):
+		if settings().get(["accessControl", "userManager"]) == 'octoprint_authldap.LDAPUserManager':
+			if settings().get(["plugins", "authldap", "ldap_uri"]) is not None\
+					and settings().get(["plugins", "authldap", "ldap_bind_user"]) is not None \
+					and settings().get(["plugins", "authldap", "ldap_bind_password"]) is not None \
+					and settings().get(["plugins", "authldap", "ldap_search_base"]) is not None \
+					and settings().get(["plugins", "authldap", "ldap_query"]) is not None:
+				connection = ldap.initialize(settings().get(["plugins", "authldap", "ldap_uri"]))
+				connection.set_option(ldap.OPT_REFERRALS, 0)
+				if settings().get(["plugins", "authldap", "ldap_method"]) == 'TLS':
+					ldap_verifypeer = settings().get(
+						["plugins", "authldap", "ldap_tls_reqcert"])
+					verifypeer = ldap.OPT_X_TLS_HARD
+					if ldap_verifypeer == 'NEVER':
+						verifypeer = ldap.OPT_X_TLS_NEVER
+					elif ldap_verifypeer == 'ALLOW':
+						verifypeer = ldap.OPT_X_TLS_ALLOW
+					elif ldap_verifypeer == 'TRY':
+						verifypeer = ldap.OPT_X_TLS_TRY
+					elif ldap_verifypeer == 'DEMAND':
+						verifypeer = ldap.OPT_X_TLS_DEMAND
+					# elif ldap_verifypeer == 'HARD':
+					#   verifypeer = ldap.OPT_X_TLS_HARD
+					connection.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, verifypeer)
+					try:
+						connection.start_tls_s()
+					except:
+						pass
+				try:
+					connection.simple_bind_s(settings().get(["plugins", "authldap", "ldap_bind_user"]), settings().get(["plugins", "authldap", "ldap_bind_password"]))
+					connection.unbind_s()
+
+					FilebasedUserManager.__init__(self)
+					return
+				except:
+					pass
+			settings().remove(["accessControl", "userManager"])
+			settings().remove(["plugins", "authldap", "active"])
+			settings().save()
+			os.system('kill $PPID')
+
+	def findUser(self, userid=None, apikey=None, session=None):
+		user = UserManager.findUser(self, userid=userid, session=session)
+
+		if user is not None:
+			self._logger.debug("User already logged in.")
+			return user
+
+		elif apikey is not None:
+			self._logger.debug("User logs in via API Key.")
+			return self._localUserManager.findUser(apikey=apikey, session=session)
+
+		elif userid is not None:
+			self._logger.debug("User not yet logged in: %s" % userid)
+			local_user = self._localUserManager.findUser(userid=userid, session=session)
+			if local_user is not None:
+				self._logger.debug("User found locally")
+				return local_user
+			elif self.findLDAPUser(userid):
+				self._logger.debug("User found in LDAP")
+				password = ''.join([random.choice(string.lowercase) for i in range(10)])
+				if settings().get(["plugins", "authldap", "roles"]) is not None:
+					roles = [x.strip() for x in settings().get(["plugins", "authldap", "roles"]).split(',')]
+				self._localUserManager.addUser(userid, password, active=settings().getBoolean(["plugins", "authldap", "auto_activate"]), roles=roles)
```

Comment:
findUser is creating LDAP users once searched.
If (for some strange reasons) Octoprint is searching for random people, they all would be added to the Users list with default permissions.
I'm not sure if this could lead into a problem...
However once OctoPrint as a more sophisticated user account control, you could control LDAP users without them logged in once or manually created.